docs/topics/request-queues.rst

```diff
@@ -0,0 +1,157 @@
+.. _topics-request-queues:
+
+==============
+Request queues
+==============
+
+.. _priority-queues:
+.. setting:: SCHEDULER_PRIORITY_QUEUE
+
+Priority queues
+===============
+
+.. currentmodule:: scrapy.pqueues
+
+To decide the order in which requests are sent, Scrapy uses a queue called the
+*priority queue*.
+
+The :setting:`SCHEDULER_PRIORITY_QUEUE` setting determines the type of priority
+queue that Scrapy uses. Its default value is :class:`ScrapyPriorityQueue`.
+
+Scrapy provides the following priority queue implementations:
+
+.. autoclass:: DownloaderAwarePriorityQueue
+
+.. autoclass:: ScrapyPriorityQueue
+
+
+.. setting:: SCHEDULER_DISK_QUEUE
+.. setting:: SCHEDULER_MEMORY_QUEUE
+.. _downstream-queues:
+
+Downstream queues
+=================
+
+.. currentmodule:: scrapy.squeues
+
+A :ref:`priority queue <priority-queues>` uses queues internally to store
+pending requests. These internal queues are called *downstream queues*.
+
+The organization method used by the downstream queues, usually either
+FIFO or LIFO, affects the order in which requests are sent for requests
+that the :ref:`priority queue <priority-queues>` considers to have the same
+priority.
+
+The type of downstream queue used depends on the :setting:`JOBDIR` setting (see
+:ref:`topics-jobs`):
+
+-   If :setting:`JOBDIR` is not set (default), the memory downstream queue type
+    defined in the :setting:`SCHEDULER_MEMORY_QUEUE` setting is used, which
+    defaults to :class:`LifoMemoryQueue`.
+
+    Scrapy provides the following memory downstream queue classes:
+
+    .. autoclass:: FifoMemoryQueue
+
+    .. autoclass:: LifoMemoryQueue
+
+-   If :setting:`JOBDIR` is set, the disk downstream queue from the
+    :setting:`SCHEDULER_DISK_QUEUE` setting is used instead, which defaults to
+    :class:`PickleLifoDiskQueue`.
+
+    In addition to request order, which disk downstream queue you use can
+    affect request serialization speed (CPU usage) and size (disk usage). See
+    “Comparison with ``marshal``” in the documentation of :mod:`pickle` for
+    more information.
+
+    When request serialization fails, Scrapy falls back to using a memory
+    downstream queue.
+
+    Scrapy provides the following disk downstream queue classes:
+
+    .. autoclass:: MarshalFifoDiskQueue
+
+    .. autoclass:: MarshalLifoDiskQueue
+
+    .. autoclass:: PickleFifoDiskQueue
+
+    .. autoclass:: PickleLifoDiskQueue
+
+
+Writing your own disk downstream queue class
+--------------------------------------------
+
+If you want to define and use your own disk downstream queue implementation,
+it has to conform to the following interface:
+
+.. class:: MyExternalQueue
+
+   .. classmethod:: __init__(cls, crawler: scrapy.crawler.Crawler, key: str, state)
+
+      Return an instance of this disk downstream queue class.
```

Comment:
`__init__` is not a `@classmethod`, and it does not “return” anything.

Also, these parameters seems to be based on the custom queue class having defined a `from_crawler` method, which it does not need to define. We should cover `from_crawler` in this class documentation as an optional classmethod, similar to how it’s covered for other Scrapy components (e.g. item pipelines). But the documentation of this `__init__` method cannot assume a `from_crawler` is defined, and so it cannot assume that the `crawler` parameter is received.

Reply:
> Also, these parameters seems to be based on the custom queue class having defined a from_crawler method

`crawler` argument is used by downstream queues to get settings etc.  

> and so it cannot assume that the crawler parameter is received.

Here I didn't understand why we cannot assume argument of function isn't passed?

Reply:
> __init__ is not a @classmethod

Which mark should we use here for proper formatting?

Reply:
> `crawler` argument is used by downstream queues to get settings etc. […] Here I didn't understand why we cannot assume argument of function isn't passed?

Oh, I see now.

Usually Scrapy components do not get a crawler directly, but they are created with `create_instance` which calls the class method `from_crawler` if defined, passing the `crawler` object and all other parameters to that class method, which then calls `__init__`. I wrongly assumed that was the case here as well.

In this case, it looks like (for historic reasons?) the `crawler` object is being passed directly on the call to `create_instance`, so there is no need to define `from_crawler` to get the `crawler` object:

https://github.com/scrapy/scrapy/blob/68379197986ae3deb81a545b5fd6920ea3347094/scrapy/pqueues.py#L75-L81

So, please forget the second part of my comment completely.

> Which mark should we use here for proper formatting?

I believe something like this (classmethod → method, cls → self):

```rst
.. method:: __init__(self, crawler: scrapy.crawler.Crawler, key: str, state)
```

Reply:
Sorry, this is very old PR and I relied on a memory. Actually you are right. We need to pass a crawler we need `from_crawler` method here, as https://github.com/scrapy/scrapy/blob/0915be7300b19e9f2df2957734e2710a709bb086/scrapy/utils/misc.py#L171-L175

Reply:
`__init__` is replaced by `from_crawler` in the documentation